linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -5349,57 +5381,109 @@ private[optimizer] object OptimizerCore {
   }
 
   private class RollbackException(val trampolineId: Int,
-      val localNameAllocatorSnapshot: FreshNameAllocator.Snapshot,
+      val localNameAllocatorSnapshot: FreshNameAllocator.Snapshot[LocalName],
       val savedMutableLocalNames: Set[LocalName],
-      val labelNameAllocatorSnapshot: FreshNameAllocator.Snapshot,
+      val labelNameAllocatorSnapshot: FreshNameAllocator.Snapshot[LabelName],
       val savedStateBackupChain: List[StateBackup],
       val cont: () => TailRec[Tree]) extends ControlThrowable
 
   class OptimizeException(val myself: AbstractMethodID,
       val attemptedInlining: List[AbstractMethodID], cause: Throwable
   ) extends Exception(exceptionMsg(myself, attemptedInlining, cause), cause)
 
-  final class FreshNameAllocator[N <: String] private (
-      createNameFun: String => N,
-      private var usedNamesToNextCounter: Map[String, Int]) {
+  private abstract class FreshNameAllocator[N <: Name] private (
+      initialMap: Map[N, Int]) {
+
     import FreshNameAllocator._
 
-    def this(createNameFun: String => N) =
-      this(createNameFun, FreshNameAllocator.InitialMap)
+    private var usedNamesToNextCounter: Map[N, Int] = initialMap
 
-    def clear(): Unit = usedNamesToNextCounter = InitialMap
+    def clear(): Unit = usedNamesToNextCounter = initialMap
 
     def freshName(base: N): N = {
       if (!usedNamesToNextCounter.contains(base)) {
         usedNamesToNextCounter = usedNamesToNextCounter.updated(base, 1)
         base
       } else {
         var i = usedNamesToNextCounter(base)
-        var result = base + "$" + i
+        val baseEncoded = base.unsafeEncoded
+        var result = createNewName(baseEncoded, i)
         while (usedNamesToNextCounter.contains(result)) {
           i += 1
-          result = base + "$" + i
+          result = createNewName(baseEncoded, i)
         }
         usedNamesToNextCounter =
           usedNamesToNextCounter.updated(base, i + 1).updated(result, 1)
-        createNameFun(result)
+        result
+      }
+    }
+
+    private def createNewName(baseEncoded: Array[Byte], suffix: Int): N = {
```

Comment:
Can we just use `withSuffix` of the relevant names here? Again, I doubt that removing the branch of non-ascii changes anything for a reasonable codebase due to branch prediction.